### ISSUE TITLE
$[ assignment not noisy enough

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10288]
Migrated from [rt.perl.org#74118](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=74118) (status was 'resolved')

Searchable as RT74118$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@vigo.rous.org



$ perl \-lwe '$\[=0'
Use of assignment to $\[ is deprecated at \-e line 1\.
$ perl \-lwe '\($\[\)=\(0\)'
$

Assignment to $\[ doesn't give the expected deprecation warning if it's
part of a list assignment\.

I noticed this in the context of the Perl4\-era libraries\.  A few of
them explicitly override $\[\, in the Perl 4 manner\, by "local\($\[\) = 0;"\,
and they surprisingly don't generate a warning thereby\.  \(None of them
assign anything other than zero to it\.\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.11.5:

Configured by zefram at Sun Feb 21 21:31:04 GMT 2010.

Summary of my perl5 (revision 5 version 11 subversion 5) configuration:
   
  Platform:
    osname=linux, osvers=2.6.26-2-686, archname=i386-linux-thread-multi
    uname='linux vigo.rous.org 2.6.26-2-686 #1 smp wed feb 10 08:59:21 utc 2010 i686 gnulinux '
    config_args='-des -Darchname=i386-linux -Dcccdlflags=-fPIC -Dccdlflags=-rdynamic -Dprefix=/home/zefram/usr/perl/perl_install/perl-5.11.5-i32-f52 -Dman1ext=1 -Dman3ext=3perl -Duselargefiles -Dusethreads -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dusedevel -Ui_db'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldl -lm -lcrypt -lutil -lpthread -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.7.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/home/zefram/usr/perl/perl_install/perl-5.11.5-i32-f52/lib/5.11.5/i386-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.11.5:
    /home/zefram/usr/perl/perl_install/perl-5.11.5-i32-f52/lib/site_perl/5.11.5/i386-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.11.5-i32-f52/lib/site_perl/5.11.5
    /home/zefram/usr/perl/perl_install/perl-5.11.5-i32-f52/lib/5.11.5/i386-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.11.5-i32-f52/lib/5.11.5
    .


Environment for perl 5.11.5:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/usr/perl/perl_install/perl-5.11.5-i32-f52/bin:/home/zefram/usr/perl/util:/home/zefram/pub/i686-pc-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/local/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From james@mastros.biz

> Assignment to $\[ doesn't give the expected deprecation warning if it's
> part of a list assignment\.
Damn\.  This is my fault\, insufficient testing\.  Too late by far for
5\.12\.0\.  Possibly 5\.12\.1?  The fix should be pretty easy\, writing it
now\.

&nbsp;   \-=\- James Mastros

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Apr 08\, 2010 at 05&#8203;:48&#8203;:13PM \+0100\, James Mastros wrote&#8203;:
> > Assignment to $\[ doesn't give the expected deprecation warning if it's
> > part of a list assignment\.
> Damn\.  This is my fault\, insufficient testing\.  Too late by far for
> 5\.12\.0\.  Possibly 5\.12\.1?  The fix should be pretty easy\, writing it
> now\.

New deprecation warnings shouldn't appear in maint releases\, so I guess
that makes it for 5\.14\.

\-\- 
In my day\, we used to edit the inodes by hand\.  With magnets\.

**Comment by p5pRT** [comment]
### From james@mastros.biz

On 10 April 2010 14&#8203;:31\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> New deprecation warnings shouldn't appear in maint releases\, so I guess
> that makes it for 5\.14\.

Yeah\, I read obra's post on maint policy just after writing that\, and
even had a short conversation with him on \#p5p\.  Certianly no\-go for
5\.12\.1\.  OTOH\, I'm hoping to remove the ability to write to it
completely by 5\.14\, so the warning becomes somewhat of a moot point\.
In any case\, this becomes a conversation for \*after\* 5\.12\.0\.

&nbsp;   \-=\- James Mastros

**Comment by p5pRT** [comment]
### From @khwilliamson

James Mastros wrote&#8203;:
> On 10 April 2010 14&#8203;:31\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
>> New deprecation warnings shouldn't appear in maint releases\, so I guess
>> that makes it for 5\.14\.
> 
> Yeah\, I read obra's post on maint policy just after writing that\, and
> even had a short conversation with him on \#p5p\.  Certianly no\-go for
> 5\.12\.1\.  OTOH\, I'm hoping to remove the ability to write to it
> completely by 5\.14\, so the warning becomes somewhat of a moot point\.
> In any case\, this becomes a conversation for \*after\* 5\.12\.0\.
> 
>    \-=\- James Mastros
> 
I can't find obra's post\.  I searched p5p and found posts from obra back 
to 2002\.

**Comment by p5pRT** [comment]
### From james@mastros.biz

http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2010/04/msg158635\.html

On 10 April 2010 15&#8203;:26\, karl williamson \<public@&#8203;khwilliamson\.com> wrote&#8203;:
> I can't find obra's post\.  I searched p5p and found posts from obra back to
> 2002\.

**Comment by p5pRT** [comment]
### From @obra




On Sat\, Apr 10\, 2010 at 08&#8203;:26&#8203;:10AM \-0600\, karl williamson wrote&#8203;:
> James Mastros wrote&#8203;:
> >On 10 April 2010 14&#8203;:31\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> >>New deprecation warnings shouldn't appear in maint releases\, so I guess
> >>that makes it for 5\.14\.
> >
> >Yeah\, I read obra's post on maint policy just after writing that\, and
> >even had a short conversation with him on \#p5p\.  Certianly no\-go for
> >5\.12\.1\.  OTOH\, I'm hoping to remove the ability to write to it
> >completely by 5\.14\, so the warning becomes somewhat of a moot point\.
> >In any case\, this becomes a conversation for \*after\* 5\.12\.0\.
> >
> >   \-=\- James Mastros
> >
> I can't find obra's post\.  I searched p5p and found posts from obra
> back to 2002\.

For those playing the home game\, "obra" is my IRC nickname\.

\-Jesse

\-\-

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed by f175a6ef89290c8aae9bf1a2f98ab12aaa44292b

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
